### DISCUSSION TITLE
How to define an variable coupled AD ElasticityTensor correctly?

### DISCUSSION THREAD

**Discussion by echo1115** [idaholab/moose#22587] [Q&A Modules: Solid mechanics] [answered]
Hi all,
I want to define an variable (or aux) coupled **AD** ElasticityTensor, but it seems that most ElasticityTensor classes in MOOSE are variable (or aux) independed , which I couldn't refer to clearly.

The poissons ratio and youngs modulus in my case  are:
_poissons_ratio = 0.23 ;
youngs_modulus= 3.0e10* exp(-porosity)* Mm  , where porosity = 0.8
Mm= (1.0+0.23*F)*(1.0+1.5e-4*(T-293.15)) , where T and F are variables to be coupled._

I used much time to create such a class and experienced many compilation errors. Coincidentally, it can be compiled successfully now. However, I'm not sure if there are potential hidden errors in it that will make my future calculations wrong. So I paste my h file and C file here.

 Acturally, I'm quite uncertain about whether I should define the 
- _temp and _fluence as 'ADVariableValue ' ?
- _porosity as 'Real' ?
- _youngs_modulus and _poissons_ratio as 'ADReal' ?
- mm as 'ADReal'?

The current way i defined them was just because the compilation succeeded. Maybe someone can address my concerns. Thanks in advance.

**ADComputeVariableCoupledElasticityTensor.h** :
```
#pragma once

#include "ComputeElasticityTensorBase.h"

class ADComputeVariableCoupledElasticityTensor : public ADComputeElasticityTensorBase
{
public:
  static InputParameters validParams();

  ADComputeVariableCoupledElasticityTensor(const InputParameters & parameters);

protected:
  virtual void computeQpElasticityTensor() override;

  /*new parameters*/
  const ADVariableValue & _temp;
  const ADVariableValue & _fluence;
  const Real & _porosity;

  ADReal _youngs_modulus;
  ADReal _poissons_ratio;

  using ADComputeElasticityTensorBase::name;
  using ADComputeElasticityTensorBase::_elasticity_tensor_name;
  using ADComputeElasticityTensorBase::issueGuarantee;
  using ADComputeElasticityTensorBase::isParamValid;
  using ADComputeElasticityTensorBase::_elasticity_tensor;
  using ADComputeElasticityTensorBase::_qp;
  using ADComputeElasticityTensorBase::_effective_stiffness;
};
```
**ADComputeVariableCoupledElasticityTensor.C** :
```
#include "ADComputeVariableCoupledElasticityTensor.h"
#include "MooseVariable.h"
#include "Coupleable.h"

registerMooseObject("myApp", ADComputeVariableCoupledElasticityTensor);

InputParameters
ADComputeVariableCoupledElasticityTensor::validParams()
{
  InputParameters params = ADComputeElasticityTensorBase::validParams();
  params.addClassDescription("Compute an elasticity tensor.");

  params.addCoupledVar("temperature", 1000.0, "Coupled temperature in units of Kelvin");
  params.addCoupledVar("fluence",0.0,"fluence.");
  params.addParam<Real>("porosity",0.8," porosity");

  return params;
}

ADComputeVariableCoupledElasticityTensor::ADComputeVariableCoupledElasticityTensor(
    const InputParameters & parameters)
    : ADComputeElasticityTensorBase(parameters),

    _temp(adCoupledValue("temperature")),
    _fluence(adCoupledValue("fluence")),
    _porosity(getParam<Real>("porosity"))
{
  // all tensors created by this class are always isotropic
  issueGuarantee(_elasticity_tensor_name, Guarantee::ISOTROPIC);
}

void
ADComputeVariableCoupledElasticityTensor::computeQpElasticityTensor()
{
    ADReal mm;
    mm = (1.0+0.23*_fluence[_qp])*(1.0+1.5e-4*(_temp[_qp]-293.15));
    _youngs_modulus = 3.0e10*exp(-_porosity)*mm;
    _poissons_ratio = 0.23;
  // Assign elasticity tensor at a given quad point
  _elasticity_tensor[_qp].fillSymmetricIsotropicEandNu(_youngs_modulus, _poissons_ratio);
}
```

**Comment by GiudGiud** [comment, marked as answer]
Hello

All the variable and number types seem appropriate.
Using ADReal for _youngs_modulus correctly makes sure the derivatives in mm are not lost when computing the young s modulus. 

_poissons_ratio does not need to be an ADReal, it could be a Real. This is inconsequential though. 

Using ADVariableValue for _temp and _fluence correctly makes sure the variables carry the derivative information. 

If you are investigating issues with your Jacobian, please use the Jacobian analyzer. 
https://mooseframework.inl.gov/help/development/analyze_jacobian.html

Guillaume

**Reply by echo1115** [reply]
Thanks for your reply and suggestions! @GiudGiud
